src/main/java/org/jenkinsci/plugins/docker/commons/KeyMaterial.java

```diff
@@ -1,55 +1,69 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
 package org.jenkinsci.plugins.docker.commons;
 
 import hudson.EnvVars;
-import hudson.model.AbstractBuild;
-import org.jenkinsci.plugins.docker.commons.impl.CompositeKeyMaterial;
-import org.jenkinsci.plugins.docker.commons.impl.NullKeyMaterial;
 
-import javax.annotation.Nullable;
 import java.io.Closeable;
 import java.io.IOException;
 import java.io.Serializable;
 
 /**
- * Represents a locally extracted credentials information.
- *
- * <p>
- * Whenever you want to fork off docker directly or indirectly, use this object to set up environment variables
- * so that docker will talk to the right daemon.
- *
- * <p>
- * When you are done using the credentials, call {@link #close()} to allow sensitive information to be removed
- * from the disk.
- *
- * @author Kohsuke Kawaguchi
- * @see DockerServerEndpoint#materialize(AbstractBuild)
- * @see DockerRegistryEndpoint#materialize(AbstractBuild)
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
  */
 public abstract class KeyMaterial implements Closeable, Serializable {
+
+    private static final long serialVersionUID = 1L;
+    
+    private final EnvVars envVars;
+
+    protected KeyMaterial(EnvVars envVars) {
+        this.envVars = envVars;
+    }
+
     /**
-     * Builds the environment variables needed to be passed when docker runs, to access
+     * Get the environment variables needed to be passed when docker runs, to access
      * {@link DockerServerCredentials} that this object was created from.
      */
-    public abstract EnvVars env();
+    public EnvVars env() {
+        return envVars;
+    }
 
     /**
      * Deletes the key materials from the file system. As key materials are copied into files
-     * every time {@link KeyMaterial} is created, it must be also cleaned up each time. 
+     * every time {@link KeyMaterialFactory} is created, it must be also cleaned up each time. 
      */
     public abstract void close() throws IOException;
-
-    /**
-     * Merge two {@link KeyMaterial}s into one.
-     */
-    public KeyMaterial plus(@Nullable KeyMaterial rhs) {
-        if (rhs==null)  return this;
-        return new CompositeKeyMaterial(this,rhs);
-    }
-
+    
     /**
      * {@link KeyMaterial} that does nothing.
      */
-    public static final KeyMaterial NULL = new NullKeyMaterial();
+    public static final KeyMaterial NULL = new KeyMaterial(new EnvVars()) {
```

Comment:
Refactor into a `private static` nested class for safe serialization (or `writeReplace` itself to `NULL`).